File: src/test/java/com/timeback/ui/viewmodel/HomeViewModelTest.java
package com.timeback.ui.viewmodel;

import com.timeback.ui.domain.model.ErrorReason;
import com.timeback.ui.domain.model.ScreenState;
import com.timeback.ui.fake.FakeFeatureGateway;
import com.timeback.ui.fake.HomeViewData;
import com.timeback.ui.feature.home.HomeViewModel;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import static org.junit.jupiter.api.Assertions.*;

/**
 * HomeViewModel 단위 테스트.
 */
@ExtendWith(LiveDataTestExtension.class)
class HomeViewModelTest {

    private FakeFeatureGateway gateway;
    private HomeViewModel viewModel;

    @BeforeEach
    void setUp() {
        gateway = new FakeFeatureGateway();
    }

    @Test
    void loadHome_shouldReturnContentWithHomeViewData() throws Exception {
        viewModel = new HomeViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertNotNull(state);
        assertInstanceOf(ScreenState.Content.class, state);

        Object data = ((ScreenState.Content) state).getData();
        assertInstanceOf(HomeViewData.class, data);
        HomeViewData hvd = (HomeViewData) data;
        assertNotNull(hvd.getWasteTimeToday());
        assertNotNull(hvd.getBaselineState());
        assertFalse(hvd.getGoalProgresses().isEmpty());
    }

    @Test
    void whenOffline_shouldReturnRetryableError() throws Exception {
        gateway.isOffline = true;
        viewModel = new HomeViewModel(gateway);
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertNotNull(state);
        assertInstanceOf(ScreenState.RetryableError.class, state);
        assertEquals(ErrorReason.OFFLINE, ((ScreenState.RetryableError) state).getReason());
    }

    @Test
    void refresh_shouldReloadHome() throws Exception {
        gateway.isOffline = true;
        viewModel = new HomeViewModel(gateway);
        Thread.sleep(300);

        gateway.isOffline = false;
        viewModel.refresh();
        Thread.sleep(300);

        ScreenState state = viewModel.state.getValue();
        assertInstanceOf(ScreenState.Content.class, state);
    }
}

File: src/test/java/com/timeback/ui/viewmodel/HomeViewModelTest.kt
package com.timeback.ui.viewmodel

import com.timeback.ui.domain.model.*
import com.timeback.ui.fake.FakeFeatureGateway
import com.timeback.ui.fake.HomeViewData
import com.timeback.ui.feature.home.HomeViewModel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.*
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.*

/**
 * UI-02 HomeViewModel 단위 테스트
 * 검증: LOADING → CONTENT, BASELINE_OBSERVING 상태, refreshHome
 */
@OptIn(ExperimentalCoroutinesApi::class)
class HomeViewModelTest {

    private val testDispatcher = UnconfinedTestDispatcher()
    private lateinit var gateway: FakeFeatureGateway
    private lateinit var viewModel: HomeViewModel

    @BeforeEach
    fun setup() {
        Dispatchers.setMain(testDispatcher)
        gateway = FakeFeatureGateway()
    }

    @AfterEach
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun `baseline ready shows full CONTENT`() = runTest {
        gateway.baselineReady = true
        viewModel = HomeViewModel(gateway)
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as HomeViewData
        assertNotNull(data.recoveryRate)
        assertFalse(data.baselineState.isObserving)
    }

    @Test
    fun `baseline observing shows observing state`() = runTest {
        gateway.baselineReady = false
        viewModel = HomeViewModel(gateway)
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as HomeViewData
        assertTrue(data.baselineState.isObserving)
        assertNull(data.recoveryRate)
    }

    @Test
    fun `refreshHome reloads data`() = runTest {
        gateway.baselineReady = false
        viewModel = HomeViewModel(gateway)
        advanceUntilIdle()

        gateway.baselineReady = true
        viewModel.refreshHome()
        advanceUntilIdle()

        val state = viewModel.state.value
        assertTrue(state is ScreenState.Content<*>)
        val data = (state as ScreenState.Content<*>).data as HomeViewData
        assertFalse(data.baselineState.isObserving)
    }
}
